File: 2021/day03.py
import sys


test = """\
00100
11110
10110
10111
10101
01111
00111
11100
10000
11001
00010
01010"""

DEBUG = 'debug' in sys.argv

if 'test' in sys.argv:
    data = test.splitlines()
else:
    data = [line.rstrip() for line in open('day03.txt')]

# Is it better to keep the input as a string, or as an integer?
# It's the difference between row[i] and row & bit.

def common(data):
    num = len(data)
    bits = len(data[0])
    allone = (1 << bits) - 1
    counts = [0]*bits
    for row in data:
        for i,c in enumerate(row):
            counts[i] += int(c)
    res = 0
    for i in counts:
        res = res * 2 + (1 if i*2 >= num else 0)
    if DEBUG:
        print(num, bits, counts, res)
    return res, allone-res

def part1(data):
    most, least = common(data)
    return most * least

def part2(data):
    bits = len(data[0])

    hdata = data
    ldata = data

    for bitno in range(bits):
        bit = 1 << (bits-bitno-1)
        if len(hdata) > 1:
            most, least = common(hdata)
            crit = '1' if bit & most else '0'
            hdata = [row for row in hdata if row[bitno] == crit]

        if len(ldata) > 1:
            most, least = common(ldata)
            crit = '1' if bit & least else '0'
            ldata = [row for row in ldata if row[bitno] == crit]

        if len(hdata) == 1 and len(ldata) == 1:
            break

    print(hdata, ldata)
    return int(hdata[0],2) * int(ldata[0],2)


print("Part 1:", part1(data) )
print("Part 2:", part2(data) )

File: 2021/day03.rs
use std::env;


fn common( bits : usize, data : &Vec<u32> ) -> (u32, u32)
{
    // We need to enumerate vertically here.

    let num: usize = data.len();
    let mut counts = vec![0; bits];

    for row in data {
        for i in 0..32 {
            if row & (1 << i) > 0 {
                counts[i] += 1;
            }
        }
    }

    let mut res : u32 = 0;
    for i in (0..bits).rev()
    {
        res = res * 2 + if counts[i] * 2 >= num { 1 } else { 0 };
    }

    let allone: u32 = (1 << bits) - 1;
    (res, allone-res)
}

fn part1( bits: usize, data: &Vec<u32> ) -> u32
{
    let (most, least) = common(bits, data);
    most * least
}

fn part2( bits: usize, data: &Vec<u32> ) -> u32
{
    let mut hdata = data.to_owned();
    let mut ldata = data.to_owned();

    for bitno in 0..bits {
        let bit = 1 << (bits-bitno-1);

        if hdata.len() > 1 {
            let (most,_least) = common(bits, &hdata);
            let crit = bit & most;
            hdata = hdata
                .into_iter()
                .filter(|row| row & bit == crit)
                .collect();
        }

        if ldata.len() > 1 {
            let (_most,least) = common(bits, &ldata);
            let crit = bit & least;
            ldata = ldata
                .into_iter()
                .filter(|row| row & bit == crit)
                .collect();
        }

        if hdata.len() == 1 && ldata.len() == 1 {
            break;
        }
    }

    println!("{} {}", hdata[0], ldata[0]);

    hdata[0] * ldata[0]
}



static TEST : [&str; 12] = 
[
    "00100",
    "11110",
    "10110",
    "10111",
    "10101",
    "01111",
    "00111",
    "11100",
    "10000",
    "11001",
    "00010",
    "01010"
];

pub fn main() {
    let live : Vec<u32> = if env::args().any(|x| x == "test")
    {
        TEST
            .iter()
            .map(|x| u32::from_str_radix(x,2).unwrap() )
            .collect()
    } else {
        include_str!("day03.txt")
            .lines()
            .map(|x| u32::from_str_radix(x,2).unwrap() )
            .collect()
    };

    let binsum = live.iter().fold(0, |accum,item| accum | item ) + 1;
    let mut bits : usize = 0;
    for i in 0..32 {
        if binsum == 1<<i {
            bits = i;
            break;
        }
    }

    println!( "Pass 1: {}", part1( bits, &live ));
    println!( "Pass 2: {}", part2( bits, &live ))
}
